File: pipeline/bgpipeline/lesson_store.py
"""``data/lessons/lessons.sqlite`` from Python, with the standard library.

A separate database from data/store.sqlite on purpose: the lessons are Backgammon Galaxy's
material, so the whole data/lessons/ folder (this file, the original exports and their
pictures) is git-ignored, while store.sqlite is committed. The schema lives in
``src/lib/lesson-store-schema.ts`` and is read from there like the main store's
(``store.schema_from_ts``), so the app and the importer cannot drift apart.

Lesson rows are a copy of an export, so a re-import may delete and rewrite one set
(``delete_set``); the pictures on disk are kept and reused. None of these functions commit.
"""

from __future__ import annotations

import sqlite3
from pathlib import Path

from .store import ROOT, open_store, schema_from_ts

LESSON_SCHEMA_TS = ROOT / "src" / "lib" / "lesson-store-schema.ts"
DEFAULT_LESSONS_DIR = ROOT / "data" / "lessons"
LESSON_STORE_FILE = "lessons.sqlite"


def open_lesson_store(path: Path) -> sqlite3.Connection:
    """Open (creating the folder and the file, or upgrading) the lesson database."""
    path.parent.mkdir(parents=True, exist_ok=True)
    return open_store(path, schema_from_ts(LESSON_SCHEMA_TS))


def find_set(conn: sqlite3.Connection, site: str, site_quiz_id: str) -> int | None:
    row = conn.execute("SELECT id FROM lesson_sets WHERE site = ? AND site_quiz_id = ?", (site, site_quiz_id)).fetchone()
    return int(row[0]) if row else None


def set_images(conn: sqlite3.Connection, site_quiz_id: str) -> dict[str, tuple[str, str]]:
    """``path -> (url, md5)`` of the pictures stored for one quiz (their paths start with its id)."""
    rows = conn.execute("SELECT path, url, md5 FROM lesson_images WHERE path LIKE ?", (f"{site_quiz_id}/%",)).fetchall()
    return {r[0]: (r[1], r[2]) for r in rows}


def owner_of_problem(conn: sqlite3.Connection, problem_id: str) -> str | None:
    """Quiz id of the set that already holds this problem, if any."""
    row = conn.execute(
        "SELECT s.site_quiz_id FROM lesson_problems p JOIN lesson_sets s ON s.id = p.set_id WHERE p.problem_id = ?",
        (problem_id,),
    ).fetchone()
    return row[0] if row else None


def delete_set(conn: sqlite3.Connection, set_id: int) -> None:
    """Delete one set's rows. The schema has no foreign keys (Python's sqlite3 would not enforce
    them anyway), so the children go explicitly: choices, problems, pictures, then the set."""
    row = conn.execute("SELECT site_quiz_id FROM lesson_sets WHERE id = ?", (set_id,)).fetchone()
    if row is None:
        return
    conn.execute("DELETE FROM lesson_choices WHERE problem_id IN (SELECT problem_id FROM lesson_problems WHERE set_id = ?)", (set_id,))
    conn.execute("DELETE FROM lesson_problems WHERE set_id = ?", (set_id,))
    conn.execute("DELETE FROM lesson_images WHERE path LIKE ?", (f"{row[0]}/%",))
    conn.execute("DELETE FROM lesson_sets WHERE id = ?", (set_id,))


def insert_set(
    conn: sqlite3.Connection,
    *,
    site: str,
    site_quiz_id: str,
    name: str,
    author: str | None,
    collection: str | None,
    problem_count: int,
    file_name: str,
    file_sha256: str,
    imported_at: str,
) -> int:
    cur = conn.execute(
        "INSERT INTO lesson_sets (site, site_quiz_id, name, author, collection, problem_count, file_name, file_sha256, imported_at) "
        "VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)",
        (site, site_quiz_id, name, author, collection, problem_count, file_name, file_sha256, imported_at),
    )
    return int(cur.lastrowid)


def insert_problem(
    conn: sqlite3.Connection,
    *,
    set_id: int,
    problem_id: str,
    number: int,
    site_problem_id: str,
    kind: str,
    image: str,
    analysis: str | None,
) -> None:
    conn.execute(
        "INSERT INTO lesson_problems (set_id, problem_id, number, site_problem_id, kind, image, analysis) VALUES (?, ?, ?, ?, ?, ?, ?)",
        (set_id, problem_id, number, site_problem_id, kind, image, analysis),
    )


def insert_choice(
    conn: sqlite3.Connection,
    *,
    problem_id: str,
    number: int,
    site_choice_id: str,
    answer: str,
    description: str | None,
    loss: float | None,
    correct: bool,
    image: str | None,
) -> None:
    conn.execute(
        "INSERT INTO lesson_choices (problem_id, number, site_choice_id, answer, description, loss, correct, image) "
        "VALUES (?, ?, ?, ?, ?, ?, ?, ?)",
        (problem_id, number, site_choice_id, answer, description, loss, 1 if correct else 0, image),
    )


def insert_image(conn: sqlite3.Connection, *, path: str, url: str, md5: str, bytes: int, width: int, height: int) -> None:
    conn.execute(
        "INSERT OR REPLACE INTO lesson_images (path, url, md5, bytes, width, height) VALUES (?, ?, ?, ?, ?, ?)",
        (path, url, md5, bytes, width, height),
    )


def set_summary(conn: sqlite3.Connection, set_id: int) -> dict:
    problems = conn.execute("SELECT COUNT(*) FROM lesson_problems WHERE set_id = ?", (set_id,)).fetchone()[0]
    choices = conn.execute(
        "SELECT COUNT(*) FROM lesson_choices WHERE problem_id IN (SELECT problem_id FROM lesson_problems WHERE set_id = ?)",
        (set_id,),
    ).fetchone()[0]
    return {"problems": int(problems), "choices": int(choices)}

File: src/lib/lesson-store.ts
/**
 * data/lessons/lessons.sqlite: the Backgammon Galaxy lessons written by
 * pipeline/import_lessons.py. Server-only (node:sqlite); the app only reads it. The database,
 * the original exports and the pictures all live in data/lessons/, which is git-ignored because
 * it is Galaxy's material. Schema: lesson-store-schema.ts (also read by bgpipeline/lesson_store.py).
 */
import path from "node:path";
import type { DatabaseSync } from "node:sqlite";
import { SCHEMA_SQL, SCHEMA_VERSION } from "./lesson-store-schema";
import {
  IMAGE_FILE_RE,
  imageSrc,
  isQuizKey,
  type LessonChoice,
  type LessonImage,
  type LessonKind,
  type LessonProblem,
  type LessonSet,
  type LessonSetSummary,
} from "./lessons";
import { openStore, withReadOnlyFile, type StoreSchema } from "./store";

export const LESSON_SCHEMA: StoreSchema = { version: SCHEMA_VERSION, sql: SCHEMA_SQL };
export const LESSON_STORE_FILE = "lessons.sqlite";

export function lessonsDir(dataDir: string): string {
  return path.join(dataDir, "lessons");
}

export function lessonStorePath(dataDir: string): string {
  return path.join(lessonsDir(dataDir), LESSON_STORE_FILE);
}

export function openLessonStore(file: string, opts: { readOnly?: boolean } = {}): DatabaseSync {
  return openStore(file, { ...opts, schema: LESSON_SCHEMA });
}

type Row = Record<string, unknown>;

const text = (v: unknown): string | null => (v == null ? null : String(v));

function toImage(storedPath: unknown, md5: unknown, width: unknown, height: unknown): LessonImage | null {
  if (storedPath == null) return null;
  const src = imageSrc(String(storedPath), text(md5));
  return src ? { src, width: Number(width ?? 0), height: Number(height ?? 0) } : null;
}

/** Every imported set, in import order (the page groups and sorts them). */
export function listLessonSets(db: DatabaseSync): LessonSetSummary[] {
  const sets = db.prepare("SELECT id, site_quiz_id, name, author, collection, imported_at FROM lesson_sets ORDER BY id").all() as Row[];
  const problems = db
    .prepare("SELECT set_id, problem_id, kind, analysis IS NOT NULL AS has_analysis FROM lesson_problems ORDER BY set_id, number")
    .all() as Row[];
  const bySet = new Map<number, Row[]>();
  for (const p of problems) {
    const list = bySet.get(Number(p.set_id));
    if (list) list.push(p);
    else bySet.set(Number(p.set_id), [p]);
  }
  return sets.map((s) => {
    const list = bySet.get(Number(s.id)) ?? [];
    const checker = list.filter((p) => p.kind === "checker").length;
    return {
      key: String(s.site_quiz_id),
      name: String(s.name),
      author: text(s.author),
      collection: text(s.collection),
      problemIds: list.map((p) => String(p.problem_id)),
      checker,
      cube: list.length - checker,
      withAnalysis: list.filter((p) => Number(p.has_analysis) !== 0).length,
      importedAt: String(s.imported_at),
    };
  });
}

/** One set with its problems and choices in order, or null. */
export function getLessonSet(db: DatabaseSync, key: string): LessonSet | null {
  const s = db
    .prepare("SELECT id, site_quiz_id, name, author, collection, file_name, imported_at FROM lesson_sets WHERE site_quiz_id = ?")
    .get(key) as Row | undefined;
  if (!s) return null;
  const problemRows = db
    .prepare(
      "SELECT p.problem_id, p.number, p.kind, p.analysis, p.image, i.md5, i.width, i.height FROM lesson_problems p " +
        "LEFT JOIN lesson_images i ON i.path = p.image WHERE p.set_id = ? ORDER BY p.number",
    )
    .all(Number(s.id)) as Row[];
  const choiceRows = db
    .prepare(
      "SELECT c.problem_id, c.number, c.site_choice_id, c.answer, c.description, c.loss, c.correct, c.image, i.md5, i.width, i.height " +
        "FROM lesson_choices c JOIN lesson_problems p ON p.problem_id = c.problem_id LEFT JOIN lesson_images i ON i.path = c.image " +
        "WHERE p.set_id = ? ORDER BY p.number, c.number",
    )
    .all(Number(s.id)) as Row[];
  const choices = new Map<string, LessonChoice[]>();
  for (const c of choiceRows) {
    const choice: LessonChoice = {
      id: String(c.site_choice_id),
      number: Number(c.number),
      answer: String(c.answer),
      description: text(c.description),
      loss: c.loss == null ? null : Number(c.loss),
      correct: Number(c.correct) !== 0,
      image: toImage(c.image, c.md5, c.width, c.height),
    };
    const list = choices.get(String(c.problem_id));
    if (list) list.push(choice);
    else choices.set(String(c.problem_id), [choice]);
  }
  const problems: LessonProblem[] = problemRows.map((p) => ({
    id: String(p.problem_id),
    number: Number(p.number),
    kind: (p.kind === "cube" ? "cube" : "checker") as LessonKind,
    image: toImage(p.image, p.md5, p.width, p.height),
    analysis: text(p.analysis),
    choices: choices.get(String(p.problem_id)) ?? [],
  }));
  const checker = problems.filter((p) => p.kind === "checker").length;
  return {
    key: String(s.site_quiz_id),
    name: String(s.name),
    author: text(s.author),
    collection: text(s.collection),
    problemIds: problems.map((p) => p.id),
    checker,
    cube: problems.length - checker,
    withAnalysis: problems.filter((p) => p.analysis).length,
    importedAt: String(s.imported_at),
    fileName: String(s.file_name),
    problems,
  };
}

/** Every imported set, or [] before the first import. */
export function readLessonSets(dataDir: string): LessonSetSummary[] {
  return withReadOnlyFile(lessonStorePath(dataDir), LESSON_SCHEMA, [], listLessonSets);
}

export function readLessonSet(dataDir: string, key: string): LessonSet | null {
  if (!isQuizKey(key)) return null;
  return withReadOnlyFile(lessonStorePath(dataDir), LESSON_SCHEMA, null, (db) => getLessonSet(db, key));
}

/** The file behind /api/lessons/images/<quiz id>/<file>, or null for any name the importer
 * does not write (so nothing outside data/lessons/<quiz id>/images/ can be reached). */
export function lessonImagePath(dataDir: string, quizId: string, file: string): string | null {
  if (!isQuizKey(quizId) || !IMAGE_FILE_RE.test(file)) return null;
  const root = path.resolve(lessonsDir(dataDir));
  const full = path.resolve(root, quizId, "images", file);
  return full.startsWith(root + path.sep) ? full : null;
}
